sql/sql_base.cc

```diff
@@ -9526,7 +9532,38 @@ bool mysql_rm_tmp_tables(void) {
 
 void tdc_flush_unused_tables() {
   table_cache_manager.lock_all_and_tdc();
-  table_cache_manager.free_all_unused_tables();
+
+  if (flush_only_old_table_cache_entries) {
+    DBUG_EXECUTE_IF("tdc_flush_unused_tables", {
+      static constexpr char act[] = "now SIGNAL signal.enabled";
+      DBUG_ASSERT(!debug_sync_set_action(current_thd, STRING_WITH_LEN(act)));
+    };);
+
+    const time_point flush_cutpoint =
+        std::chrono::system_clock::now() - std::chrono::seconds(flush_time);
+    table_cache_manager.free_old_unused_tables(flush_cutpoint);
+
+    /* Free table shares which were not freed implicitly by loop above. */
+    std::vector<TABLE_SHARE *> old_entries;
+    for (TABLE_SHARE *s = oldest_unused_share; s->next; s = s->next) {
+      if (should_be_evicted(s->last_accessed, flush_cutpoint)) {
+        old_entries.push_back(s);
```

Comment:
Do we really need to save table share pointers to be evicted in a separate container.
Can they be deleted in-place?
```suggestion
        table_def_cache->erase(to_string(s->table_cache_key));
```